### PR TITLE
Improve processors

### LINKED ISSUE #344: Unit test and validate the output of the match cost formula
I implemented this but never got around to setting up unit tests for it. Should definitely get done at some point because I am terrible at math and it took Cytusine explaining it to me like a 2 year old and 2 hours of my brain at max capacity (it still looks kinda jank)

- [x] Unit test coverage and output validation of `DataWorkerService.Utilities.MatchCostCalculator`

### LINKED ISSUE #361: Add configurable osu! and osu!track API ratelimits
Add configuration fields for osu! and osu!track API ratelimits / window timings

### PR DESCRIPTION
**Pull request by myssto**
Closes #344
Closes #361

Addresses all existing reliability and consistency concerns with the processors in the DataWorker.

- Add a Verification processor step that checks for concrete verification and rejection
- Greatly improved reliability of the `MatchHeadToHeadCheck` (TeamVs Converter)
- Use multithreading for non-blocking operations in processors
- Fix and unit test the `MatchCostCalculator`
- Unit testing for stats processors
- Support for configuring the rate limits of the `OsuClient`
- Improve the efficiency of the `OsuApiDataParserService` with caching and less redundant database interactions

### CODE REVIEW

File: DataWorkerService.Tests/Stats/GameStatsProcessorTests.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        Assert.Equal(expectedWinningTeam, result.WinnerTeam);
+        Assert.Equal(expectedLosingTeam, result.LoserTeam);
+        Assert.Equal([1, 2, 3], result.WinnerRoster);
+        Assert.Equal([4, 5, 6], result.LoserRoster);
+        Assert.Equal(1500, result.WinnerScore);
+        Assert.Equal(600, result.LoserScore);
+    }
+
+    [Fact]
+    public async void Processor_Integrated_BehavesAsExpected()
```
Comment (hburn7) [inline DataWorkerService.Tests/Stats/GameStatsProcessorTests.cs:64]:
```suggestion
    public async void Processor_Integrated_ProcessesGameSuccessfully()
```

File: DataWorkerService/Services/Implementations/OsuApiDataParserService.cs [hunk 2]
```diff
@@ -31,21 +34,110 @@ public async Task ParseMatchAsync(Match match, MultiplayerMatch apiMatch)
...
 
         match.EndTime = DetermineMatchEndTime(apiMatch) ?? default;
 
         logger.LogDebug("Finished parsing match [Id: {Id} | osu! Id: {OsuId}]", match.Id, match.OsuId);
     }
 
-    public async Task<IEnumerable<Game>> ParseGamesAsync(MultiplayerMatch apiMatch)
+    public async Task LoadPlayersAsync(IEnumerable<MatchUser> apiPlayers)
+    {
+        var eApiPlayers = apiPlayers.ToList();
```
Comment (hburn7) [inline DataWorkerService/Services/Implementations/OsuApiDataParserService.cs:51]: `eApiPlayers` is a really ugly var name

File: DataWorkerService/Services/Implementations/OsuApiDataParserService.cs [hunk 3]
```diff
@@ -31,21 +34,110 @@ public async Task ParseMatchAsync(Match match, MultiplayerMatch apiMatch)
...
+                playersRepository.Add(player);
+            }
+
+            _playersCache.Add(player);
+        }
+    }
+
+    public async Task LoadBeatmapsAsync(IEnumerable<MultiplayerGame?> apiGames)
+    {
+        var eGames = apiGames.ToList();
```
Comment (hburn7) [inline DataWorkerService/Services/Implementations/OsuApiDataParserService.cs:88]: ew

File: DataWorkerService/Utilities/MatchCostCalculator.cs [hunk 4]
```diff
@@ -14,49 +14,53 @@ public static class MatchCostCalculator
...
+    /// Only <see cref="GameScore"/>s with a <see cref="VerificationStatus"/> of <see cref="VerificationStatus.Verified"/>
+    /// and a <see cref="ScoreProcessingStatus"/> of <see cref="ScoreProcessingStatus.Done"/> are considered
+    /// </remarks>
     public static IDictionary<int, double> CalculateOtrMatchCosts(IEnumerable<Game> games)
     {
-        IEnumerable<Game> enumerableGames = games as Game[] ?? games.ToArray();
+        IEnumerable<Game> eGames = games.ToList();
 
-        var zScores = enumerableGames
+        var zScores = eGames
```
Comment (hburn7) [inline DataWorkerService/Utilities/MatchCostCalculator.cs:26]: var name

**Review by hburn7 — CHANGES_REQUESTED** [review]

File: DataWorkerService/Services/Implementations/OsuApiDataParserService.cs (hunk 3 above)
Comment (myssto) [reply DataWorkerService/Services/Implementations/OsuApiDataParserService.cs:88]: youre ew

**Review by hburn7 — APPROVED** [review]

**Pull request merged**